src/core/linalg/src/dense/4C_linalg_tensor.hpp

```diff
@@ -0,0 +1,366 @@
+       * @brief Access to the underlying readonly container (std::array or std::span, depending on
+       * whether it is owning or a view)
+       *
+       * @return Number*
+       */
+      [[nodiscard]] const ContainerType& container() const { return data_; }
+
+      /*!
+       * @brief Indexing operator to access individual values of the tensor (without bound checks)
+       *
+       * @param i
+       * @return Number&
+       */
+      [[nodiscard]] Number& operator()(decltype(n)... i)
+      {
+        return data_[get_flat_index<OrderType::column_major, TensorBoundCheck::no_check, n...>(
+            i...)];
+      }
+
+      /*!
+       * @brief Indexing operator to access individual values of the tensor in readonly mode
+       * (without bound checks)
+       *
+       * @param i
+       * @return Number&
+       */
+      [[nodiscard]] const Number& operator()(decltype(n)... i) const
+      {
+        return data_[get_flat_index<OrderType::column_major, TensorBoundCheck::no_check, n...>(
+            i...)];
+      }
+
+      /*!
+       * @brief Indexing operator to access individual values of the tensor (with bound checks)
+       *
+       * @param i
+       * @return Number&
+       */
+      [[nodiscard]] Number& at(decltype(n)... i)
+      {
+        return data_[get_flat_index<OrderType::column_major, TensorBoundCheck::check, n...>(i...)];
+      }
+
+      /*!
+       * @brief Indexing operator to access individual values of the tensor in readonly mode
+       * (with bound checks)
+       *
+       * @param i
+       * @return Number&
+       */
+      [[nodiscard]] const Number& at(decltype(n)... i) const
+      {
+        return data_[get_flat_index<OrderType::column_major, TensorBoundCheck::check, n...>(i...)];
+      }
+
+      [[nodiscard]] static constexpr std::size_t rank() { return rank_; }
+      [[nodiscard]] static constexpr std::size_t size() { return size_; }
+
+      /*!
+       * @brief Returns the dimension of the i-th axis of the tensor
+       *
+       * @tparam i
+       */
+      template <std::size_t i>
+        requires(i < rank())
+      [[nodiscard]] static constexpr auto size_of()
+      {
+        return std::get<i>(std::make_tuple(n...));
+      }
+
+      /*!
+       * @brief Returns a std::tuple of the shape of the tensor
+       *
+       * @return constexpr auto
+       */
+      [[nodiscard]] static constexpr auto shape() { return std::make_tuple(n...); }
+
+      /*!
+       * @brief Fills the tensor with the given value
+       *
+       * @param value
+       */
+      constexpr void fill(const Number& value) { std::ranges::fill(data_, value); }
+
+      /*!
+       * @brief Add another tensor onto this tensor
+       *
+       * @tparam OtherTensor
+       */
+      template <typename OtherTensor>
+        requires(std::is_same_v<typename OtherTensor::shape_type, typename OtherTensor::shape_type>)
+      constexpr Tensor<Number, storage_type, n...>& operator+=(const OtherTensor& B)
```

Comment:
I tried to use this in a `consteval` function. It does not work yet since `DenseFunctions::update` is not `constexpr`. Could be addressed at a later point.